omniledger/contracts/coins.go

```diff
@@ -0,0 +1,197 @@
+//    instanceID
+//  - fetch takes "coins" out of the account and returns it as an output
+//    parameter for the next instruction to interpret.
+//  - store puts the coins given to the instance back into the account.
+// You can only delete a contractCoin instance if the account is empty.
+func ContractCoin(cdb omniledger.CollectionView, inst omniledger.Instruction, c []omniledger.Coin) (sc []omniledger.StateChange, cOut []omniledger.Coin, err error) {
+	cOut = c
+	switch inst.GetType() {
+	case omniledger.SpawnType:
+		// Spawn creates a new coin account as a separate instance. The subID is
+		// taken from the hash of the instruction.
+		ca := omniledger.InstanceID{
+			DarcID: inst.InstanceID.DarcID,
+			SubID:  omniledger.NewSubID(inst.Hash()),
+		}
+		log.Lvlf3("Spawing coin to %x", ca.Slice())
+		sc = []omniledger.StateChange{
+			omniledger.NewStateChange(omniledger.Create, ca, ContractCoinID, make([]byte, 8)),
+		}
+		return
+	case omniledger.InvokeType:
+		// Invoke is one of "mint", "transfer", "fetch", or "store".
+		var value []byte
+		value, _, err = cdb.GetValues(inst.InstanceID.Slice())
+		if err != nil {
+			return
+		}
+		coinsCurrent := newSafeUint64(value)
+		var coinsArg uint64
+
+		if inst.Invoke.Command != "store" {
+			coinsBuf := inst.Invoke.Args.Search("coins")
+			if coinsBuf == nil {
+				err = errors.New("argument \"coins\" is missing")
+				return
+			}
+			coinsArg = binary.LittleEndian.Uint64(coinsBuf)
+		}
+		switch inst.Invoke.Command {
+		case "mint":
+			// mint simply adds this amount of coins to the account.
+			log.Lvl2("minting", coinsArg)
+			coinsCurrent, err = coinsCurrent.add(coinsArg)
+			if err != nil {
+				return
+			}
+		case "transfer":
+			// transfer sends a given amount of coins to another account.
+			coinsCurrent, err = coinsCurrent.sub(coinsArg)
+			if err != nil {
+				return
+			}
+
+			target := inst.Invoke.Args.Search("destination")
+			var (
+				v   []byte
+				cid string
+			)
+			v, cid, err = cdb.GetValues(target)
+			if err == nil && cid != ContractCoinID {
+				err = errors.New("destination is not a coin contract")
+			}
+			if err != nil {
+				return
+			}
+
+			targetCoin := newSafeUint64(v)
+			targetCoin, err = targetCoin.add(coinsArg)
+			if err != nil {
+				return
+			}
+			var w bytes.Buffer
+			binary.Write(&w, binary.LittleEndian, targetCoin)
+
+			log.Lvlf3("transferring %d to %x", coinsArg, target)
+			sc = append(sc, omniledger.NewStateChange(omniledger.Update, omniledger.NewInstanceID(target),
+				ContractCoinID, w.Bytes()))
+		case "fetch":
+			// fetch removes coins from the account and passes it on to the next
+			// instruction.
+			if coinsArg > uint64(coinsCurrent) {
+				err = errors.New("not enough coins in instance")
+				return
+			}
+			coinsCurrent, err = coinsCurrent.sub(coinsArg)
+			if err != nil {
+				return
+			}
+			cOut = append(cOut, omniledger.Coin{Name: CoinName, Value: coinsArg})
```

Comment:
I don't understand why coins needs a name. From my understanding, we should enforce that store/fetch work only in the same instance (not sure how to do it).